### PR TITLE
Add iftop support to status_graph.php

### PR DESCRIPTION
**Pull request by joshuasign**
- [ ] Redmine Issue: https://redmine.pfsense.org/issues/3334
- [x] Ready for review

### CODE REVIEW

**Comment by joshuasign** [comment]
changes done.

**Comment by bimmerdriver** [comment]
I installed the four patches (in order) with no apparent errors. If I run iftop from the command line, there is ipv4 and ipv6 traffic, but no hosts are showing up beside the graph at all (for either protocol) if I select the experimental mode. The regular mode still shows ipv4 traffic.

**Comment by joshuasign** [comment]
@bimmerdriver : four patches ? I didn't provide any patch for this final version of this PR.

you can get one from : https://patch-diff.githubusercontent.com/raw/pfsense/pfsense/pull/4039.patch
but take care about permissions on /usr/local/bin/iftop_parser.sh, it seems to be wrong in the patch url.

But the commit seems good : https://github.com/pfsense/pfsense/pull/4039/commits/3875a3d50198f5e747b1ee76e1a728aaeaa6f644

**Comment by bimmerdriver** [comment]
@joshuasign When I said "four patches", I was referring to the four individual commits in this PR. When I installed them in order, from first to last, there were no errors, but the feature did not work. (I'm using the latest development snapshot.) I tried using the number you posted above, but it would not install. I backed out of all of the previous patches with out any errors, so hopefully my system is not messed up. If you create a consolidated patch against the latest development snapshot, I will try to install and test it.

**Comment by jim-p** [comment]
You shouldn't need to apply individual commits. Just add `.diff` to the PR URL and use that. For example, this PR is `https://github.com/pfsense/pfsense/pull/4039` so the patch to apply is `https://github.com/pfsense/pfsense/pull/4039.diff`. The format of ending in `.patch` is different and does not always apply cleanly as it's not intended for that purpose.

**Comment by joshuasign** [comment]
Jim, i notice that the patch url doesn't take care about permissions i give to the script iftop_parser.sh
am'i wrong ?

**Comment by jim-p** [comment]
It does in both places, but `.diff` does all changes at once, and `.patch` does them in commit order. The permission change is there way down far in the patch. The diff is what you want to apply for testing, the patch is for feeding into another git repo to preserve the commit history.

tl;dr use .diff not .patch

**Comment by bimmerdriver** [comment]
@jim-p Thank you for the reply. I installed the patch as you suggested. The result is the same as when I installed it in pieces. It installs cleanly, but it doesn't work. Using rate, I can see ipv4 traffic. Using iftop, I can't see either ipv4 or ipv6 traffic. If I run iftop from the command line, it shows both ipv4 and ipv6 traffic.

**Comment by bimmerdriver** [comment]
Here is some output from iftop:

[2.4.5-DEVELOPMENT][admin@pfSense.localdomain]/root: iftop -nNb -i hn0 -s 1 -o 2s -t
interface: hn0
IP address is: 192.168.1.1
MAC address is: 00:15:5d:5c:ffffffe2:2d
Listening on hn0
^[[A   # Host name (port/service if enabled)            last 2s   last 10s   last 40s cumulative
--------------------------------------------------------------------------------------------
   1 172.93.16.154                            =>     4.08Mb     4.08Mb     4.08Mb     1.02MB
     192.168.1.103                            <=     83.0Kb     83.0Kb     83.0Kb     20.7KB
   2 2001:569:74b0:6000:215:5dff:fe5c:e22d    =>     1.00Kb     1.00Kb     1.00Kb       256B
     2001:569:74b0:6000::2000                 <=       680b       680b       680b       170B
   3 192.168.1.1                              =>       852b       852b       852b       213B
     192.168.1.103                            <=       580b       580b       580b       145B
   4 2610:160:11:18::199                      =>       480b       480b       480b       120B
     2001:569:74b0:6000:8cc7:33cc:dcfd:9a39   <=       720b       720b       720b       180B
   5 2606:6d00:0:192::154                     =>         0b         0b         0b         0B
     2001:569:74b0:6000:8cc7:33cc:dcfd:9a39   <=       288b       288b       288b        72B
--------------------------------------------------------------------------------------------
Total send rate:                                     4.09Mb     4.09Mb     4.09Mb
Total receive rate:                                  85.2Kb     85.2Kb     85.2Kb
Total send and receive rate:                         4.17Mb     4.17Mb     4.17Mb
--------------------------------------------------------------------------------------------
Peak rate (sent/received/total):                     4.09Mb     85.2Kb     4.17Mb
Cumulative (sent/received/total):                    1.02MB     21.3KB     1.04MB
============================================================================================

[2.4.5-DEVELOPMENT][admin@pfSense.localdomain]/root: iftop -Nb -i hn0 -s 1 -o 2s -t
interface: hn0
IP address is: 192.168.1.1
MAC address is: 00:15:5d:5c:ffffffe2:2d
Listening on hn0
   # Host name (port/service if enabled)            last 2s   last 10s   last 40s cumulative
--------------------------------------------------------------------------------------------
   1 2606:6d00:0:192::154                     =>     7.91Mb     7.91Mb     7.91Mb     1.98MB
     node-1w7jr9ql7zdvv58pbadhcrvw9.ipv6.tel  <=      147Kb      147Kb      147Kb     36.7KB
   2 2610:160:11:18::199                      =>     6.07Kb     6.07Kb     6.07Kb     1.52KB
     node-1w7jr9ql7zdvv58pbadhcrvw9.ipv6.tel  <=     4.74Kb     4.74Kb     4.74Kb     1.19KB
   3 pfSense.localdomain                      =>       480b       480b       480b       120B
     192.168.1.103                            <=       320b       320b       320b        80B
   4 172-93-16-154.ebox.net                   =>       160b       160b       160b        40B
     192.168.1.103                            <=       160b       160b       160b        40B
--------------------------------------------------------------------------------------------
Total send rate:                                     7.92Mb     7.92Mb     7.92Mb
Total receive rate:                                   152Kb      152Kb      152Kb
Total send and receive rate:                         8.07Mb     8.07Mb     8.07Mb
--------------------------------------------------------------------------------------------
Peak rate (sent/received/total):                     7.92Mb      152Kb     8.07Mb
Cumulative (sent/received/total):                    1.98MB     38.0KB     2.02MB
============================================================================================

**Comment by jim-p** [comment]
@bimmerdriver Check `/usr/local/bin/iftop_parser.sh`, ensure it is mode 755. Try running `/usr/local/bin/iftop_parser.sh hn0` a few times and see if any traffic appears.

**Comment by jim-p** [comment]
Also try setting the **Filter** option on the traffic graph to *All*.

**Comment by bimmerdriver** [comment]
The permission on the script is fine. Here is some output:

[2.4.5-DEVELOPMENT][admin@pfSense.localdomain]/usr/local/bin: iftop_parser.sh hn0
/usr/bin/awk: extra } at source line 42 source file /usr/local/bin/iftop_parse.awk
 context is
         >>> } <<<
/usr/bin/awk: syntax error at source line 43 source file /usr/local/bin/iftop_parse.awk
        2 extra }'s
/usr/bin/awk: bailing out at source line 113
/usr/bin/awk: extra } at source line 42 source file /usr/local/bin/iftop_parse.awk
 context is
         >>> } <<<
/usr/bin/awk: syntax error at source line 43 source file /usr/local/bin/iftop_parse.awk
        2 extra }'s
/usr/bin/awk: bailing out at source line 113
/usr/bin/awk: extra } at source line 42 source file /usr/local/bin/iftop_parse.awk
 context is
         >>> } <<<
/usr/bin/awk: syntax error at source line 43 source file /usr/local/bin/iftop_parse.awk
        2 extra }'s
/usr/bin/awk: bailing out at source line 113
[2.4.5-DEVELOPMENT][admin@pfSense.localdomain]/usr/local/bin:

**Comment by bimmerdriver** [comment]
[2.4.5-DEVELOPMENT][admin@pfSense.localdomain]/usr/local/bin: ls -l iftop_parser.sh
-rwxr-xr-x  1 root  wheel  4228 Jan 26 10:40 iftop_parser.sh

**Comment by bimmerdriver** [comment]
[2.4.5-DEVELOPMENT][admin@pfSense.localdomain]/usr/local/bin: wc iftop_parser.sh
     153     540    4227 iftop_parser.sh

**Comment by bimmerdriver** [comment]
@joshuasign The crlf is whatever was in the file in the patch. I noticed iftop_parse.awk and .orig and iftop_parser.sh and .orig. If I revert the patch, should these files still be installed? If not, I will remove them and reinstall the patch. Maybe it got corrupted somehow.

**Comment by bimmerdriver** [comment]
Is anything happening with this?

**Comment by jim-p** [comment]
I've been using it for months on my edge firewall and it's been OK, but still has some quirks. I'm not sure if it's iftop itself (likely) or somewhere else in the code. I haven't tracked any of the issues I've seen down to a specific cause yet. I'm inclined to pull it into 2.5.0 at this point but maybe not make it the default yet.

**Comment by jim-p** [comment]
I don't have any notes with what they were handy. IIRC when you switch some settings, occasionally it just doesn't display anything until you refresh the page and try again. I think it also didn't work correctly on some type of interface, but I can't remember which right at the moment. Also there was something with the traffic total amounts being off, but it may be like you said, they come from different sources, but they still seemed much lower overall than they should be even with sustained transfers that keep the rates constant.

File: src/usr/local/bin/iftop_parser.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                $RM $pid_file
+                $RM $cache_file 2>> /dev/null
+        else
+                if [ -s $cache_file ]; then
+                        $CAT $cache_file
+                fi
+        fi
+else
+        echo -n $$ > $pid_file
+        $IFTOP -nNb -i $1 -s 1 -o 2s -t 2>> /dev/null | $AWK -f $awk_script > ${cache_file}.tmp
```
Comment (joshuasign) [inline src/usr/local/bin/iftop_parser.sh]: mistake in argument -s 1
should be -s 2

**Comment by joshuasign** [comment]
@jim-p ok, i found the problem about  rates : 
there were a mistake in [iftop_parser.sh line49](https://github.com/pfsense/pfsense/pull/4039/files#diff-e2f25fba04e497cb17ef4eaf244de3d1R49)
Please try with -s 3, i'll post a commit latter if it's ok for you.
I do tests with -s 4, but values are exactly the sames.

This is because iftop need at least 2 seconds to check rates.
But with -s 2 it seems to be a little too short, so -s 3 may works better.

I compare with the return of an iftop command without the -s argument.
And the values are the sames : [iftop console exemple](https://image.noelshack.com/fichiers/2019/14/2/1554157298-exemple-iftop-console.png
)

Iftop mode with -s 3 gives the same value than the chart, and sometimes a little less :
[iftop exemple 1](https://image.noelshack.com/fichiers/2019/14/2/1554157298-exemple-iftop-1.png)
[iftop exemple 2](https://image.noelshack.com/fichiers/2019/14/2/1554157298-exemple-iftop-2.png)

I also test the original "rates" mode : it gives not the same values than the chart too.
But it always gives a constant value : 
[rates mode exemple](https://image.noelshack.com/fichiers/2019/14/2/1554157298-exemple-rate.png)

Note : on iftop mode, values cant be refresh more than one time every 3 seconds.

**Comment by jim-p** [comment]
I tried with a few different values and the results were more or less the same. Though it also agrees with what iftop shows at the CLI in the curses output. Maybe the graph data is wrong because it shows a value much higher than any of the totals. For example right now it shows 1Mbit/s on LAN but the totals are nowhere near that in rate, so that's probably not an issue worth pursuing if you are happy with `-s 3`.

The main problem I have now is that sometimes when I change from one interface to another there is no output at all. If I flip an option, say to another interface and then back or change the filtering, it may start. It's not clear yet what is happening there. Unfortunately, it's not easy to reproduce or predict. It may happen twice in a row, it may not happen for 10 changes.

**Comment by bimmerdriver** [comment]
I just reapplied the patch and I also reinstalled the iftop package. The option to use iftop is available, but no hosts are showing up. I'm running the latest 2.5.0 snapshot. Any suggestions?

**Comment by jim-p** [comment]
Works for me here on 2.5.0. You do still need to run `chmod a+x /usr/local/bin/iftop_parser.sh` after applying the patch on a system for the first time.

**Comment by bimmerdriver** [comment]
@jim-p I did that and it still doesn't work. Something is probably messed up with my system, but not sure what. What URL should I use for the patch? Is there a way to restore the configuration?

**Comment by jim-p** [comment]
Just revert the patch and you'll be fine. The URL I used is the URL for the PR plus `.diff`, so: https://github.com/pfsense/pfsense/pull/4039.diff

**Comment by bimmerdriver** [comment]
@jim-p That made no difference. The iftop option is available and there is traffic on the graph, but no hosts are being shown. It was working before, but it's been broken for a while. Something must be effed up on my system. Is there a way to restore the configuration?

File: src/etc/inc/util.inc [hunk 2]
```diff
@@ -1962,6 +1962,23 @@ function format_number($num, $precision = 3) {
...
+    } else if ( strpos($num,"G") !== false ) {
+        $num = str_replace("G","",$num) * 1000 * 1000 * 1000;
+    } else if ( strpos($num,"M") !== false ) {
+        $num = str_replace("M","",$num) * 1000 * 1000;
+    } else if ( strpos($num,"K") !== false ) {
+        $num = str_replace("K","",$num) * 1000;
+    }
+    
+    return $num;
+}
```
Comment (rbgarga) [inline src/etc/inc/util.inc]: Please fix indentation here. We use hard TAB with size 8

File: src/usr/local/www/bandwidth_by_ip.php [hunk 3]
```diff
@@ -119,10 +181,10 @@
 	// echo $bandwidthinfo;
 	$emptyinfocounter = 1;
 	if ($bandwidthinfo != "") {
-		$infoarray = explode (":", $bandwidthinfo);
+		$infoarray = explode (";", $bandwidthinfo);
 		if (($filter == "all") ||
-		    (($filter == "local") && (ip_in_subnet($infoarray[0], $intsubnet))) ||
-		    (($filter == "remote") && (!ip_in_subnet($infoarray[0], $intsubnet)))) {
+			(($filter == "local") && ((ip_in_subnet($infoarray[0], $intsubnet)) || (ip_in_subnet($infoarray[0], $intsubnet6)))) ||
+			(($filter == "remote") && ((!ip_in_subnet($infoarray[0], $intsubnet)) || (!ip_in_subnet($infoarray[0], $intsubnet6))))) {
```
Comment (rbgarga) [inline src/usr/local/www/bandwidth_by_ip.php]: Multiline statements should use 4 spaces to indent, as it was before the change. Please leave it as it is

**Comment by rbgarga** [comment]
Superseded by #4181
